bookserver/routers/discuss.py

```diff
@@ -21,19 +70,162 @@
-    await websocket.accept()
-    while True:
-        data = await websocket.receive_text()
-        await websocket.send_text(f"Message sent was {data}")
+class ConnectionManager:
+    def __init__(self):
+        self.active_connections: Dict[str, WebSocket] = {}
+
+    async def connect(self, user: str, websocket: WebSocket):
+        await websocket.accept()
+        self.active_connections[user] = websocket
+
+    def disconnect(self, sockid: str):
+        del self.active_connections[sockid]
+
+    async def send_personal_message(
+        self,
+        receiver: str,
+        message: PeerMessage,
+    ):
+        to = receiver
+        if to in self.active_connections:
+            try:
+                rslogger.debug(f"sending PM to {to} on {self.active_connections[to]}")
+                await self.active_connections[to].send_json(message)
+            except Exception as e:
+                rslogger.error(f"Error sending to {to} is {e}")
+                del self.active_connections[to]
+        else:
+            rslogger.error(f"{to} is not connected {self.active_connections}")
+
+    async def broadcast(self, message: str):
+        rslogger.debug(f"{self.active_connections=} {message=}")
+        to_remove = []
+        for key, connection in self.active_connections.items():
+            rslogger.debug(f"sending to {connection}")
+            res = None
+            try:
+                res = await connection.send_json(message)
+            except Exception as e:
+                rslogger.debug(f"Failed to send {e}")
+                to_remove.append(key)
+            rslogger.debug(f"result of send = {res}")
+        for key in to_remove:
+            del self.active_connections[key]
+
+
+# this is good for prototyping, but we will need to integrate with
+# Redis or a DB for production where we have multiple servers
+manager = ConnectionManager()
+local_users = set()
+
+
+async def get_cookie_or_token(
+    websocket: WebSocket,
+    access_token: Optional[str] = Cookie(None),
+    user: Optional[str] = Query(None),
+):
+    rslogger.debug(f"HELLO {access_token=} or {user=}")
+    if access_token is None and user is None:
+        await websocket.close(code=status.WS_1008_POLICY_VIOLATION)
+    return access_token or user
+
+
+# It seems that ``@router.websocket`` is much better than the documented
+# ``websocket_route``
+@router.websocket("/chat/{uname}/ws")
+async def websocket_endpoint(websocket: WebSocket, uname: str):
+    """
+    This endpoint is called to establish a websocket connection between
+    The browser and the server.  The websocket is persistent as long
+    as the process that runs this endpoint is alive.
+
+    Note: This function must return which means it must be throroughly async
+    Using a non async library like plain redis-py will not work as the subscriber
+    will block
+    """
+    rslogger.debug(f"IN WEBSOCKET {uname=}")
+    username = uname
+    local_users.add(username)
```

Comment:
You might add `global local_users` just to make it clear where this var comes from. (Not required, but helps explain.)